File: LeetCode/Medium/0279-perfect-squares/0279-perfect-squares.cpp
// class Solution {
// public:
//     int numSquares(int n) {
//         std::vector<int> dp(n + 1, n); 
//         dp[0] = 0;
//         for (int i = 1; i <= n; ++i) {
//             for (int j = 1; j * j <= i; ++j) {
//                 dp[i] = std::min(dp[i], dp[i - j * j] + 1);
//             }
//         }
//         return dp[n];
//     }
// };



class Solution {
    static bool isSquare(int x) {
        int r = (int)std::sqrt(x);
        return r * r == x;
    }
public:
    int numSquares(int n) {
        if (isSquare(n)) return 1;
        // 去掉 4 的因子
        int m = n;
        while ((m & 3) == 0) m >>= 2;
        if (m % 8 == 7) return 4;
        // 是否为两个平方和
        for (int a = 1; a * a <= n; ++a) {
            if (isSquare(n - a * a)) return 2;
        }
        return 3;
    }
};

File: LeetCode/Medium/0279-perfect-squares/0279-perfect-squares.ts
// function numSquares(n: number): number {
//     const dp = new Array<number>(n + 1).fill(n);
//     dp[0] = 0;
//     for (let i = 1; i <= n; i++) {
//         for (let j = 1; j * j <= i; j++) {
//             dp[i] = Math.min(dp[i], dp[i - j * j] + 1);
//         }
//     }
//     return dp[n];
// };


function numSquares(n: number): number {
  const isSquare = (x: number) => {
    const r = Math.trunc(Math.sqrt(x));
    return r * r === x;
  };

  if (isSquare(n)) return 1;

  let m = n;
  while ((m & 3) === 0) m >>= 2; // 去掉4因子
  if (m % 8 === 7) return 4;

  for (let a = 1; a * a <= n; a++) {
    if (isSquare(n - a * a)) return 2;
  }
  return 3;
}